Airspeed preflight checks update

sfuhrer opened the pull request:
Some updates to the airspeed preflight checks.

**Describe problem solved by this pull request**
- airspeed validity information given by the airspeed selector/validator is checked for preflight checks
- user can set maximal allowed airspeed during arming (new parameter ASPD_MAX_ARM)

**Describe your solution**
- ~~ listen to airspeed_validated instead of airspeed in the airspeed preflight checks~~
- ~~don't allow arming if airspeed_selector is down (as we then have no airspeed_validated data)~~
- ~~add confidence value to the airspeed_validated message, such that this can be checked at instant of arming (later we can also think about how to include it in the general airspeed validation checks)~~
- ~~add new parameter ASPD_MAX_ARM to set maximal allowed airspeed during arming (instead of hard-coded 4 m/s like now)~~
- ~~rename subsystem DIFFERENTIALPRESSURE to AIRSPEED as airspeed source doesn't need to be a differential pressure sensor~~


**Test data / coverage**
SITL and bench tested. 

Edit: I reduced the scope of this PR, now:
-check for valid airspeed_validated (declared valid plus updated less than 1s ago)
-add param for max airspeed for arming instead of hardcoded 4m/s

It removes the check for the airspeed.confidence, which I anyway haven't seen working properly, and instead listens to what the airspeed selector says (airspeed valid or not). We may want to add additional checks in the airspeed selector at some point for failure detection on the ground (like for detecting a stuck sensor).

RomanBapst commented on src/lib/airspeed_validator/AirspeedValidator.hpp:
```diff
@@ -124,6 +126,7 @@ class AirspeedValidator
 	float _EAS{0.0f}; ///< equivalent airspeed in m/s
 	float _TAS{0.0f}; ///< true airspeed in m/s
 	float _EAS_scale{1.0f}; ///< scale factor from IAS to EAS
+	float _airspeed_confidence{0.0f}; ///< metric form airspeed driver indicating current confidence level of the data
```
@sfuhrer Typo

sfuhrer replied:
addressed.

RomanBapst commented on src/modules/airspeed_selector/airspeed_selector_params.c:
```diff
@@ -194,3 +194,17 @@ PARAM_DEFINE_INT32(ASPD_FS_T2, 100);
 PARAM_DEFINE_FLOAT(ASPD_STALL, 10.0f);
+
+/**
+ * Maximal airspeed allowed for arming.
```
```suggestion
 * Maximum airspeed allowed for arming.
```

RomanBapst replied:
@sfuhrer Actually this seems to be correct, although it sound weird to me.

sfuhrer replied:
true, fixed.

RomanBapst commented on src/modules/airspeed_selector/airspeed_selector_params.c:
```diff
@@ -194,3 +194,17 @@ PARAM_DEFINE_INT32(ASPD_FS_T2, 100);
+ * Maximal airspeed allowed for arming.
+ *
+ * It is not possible to arm the vehicle if the current airspeed estimate (equivalent airspeed) at arming is above this value.
+ * Airspeed on ground is either comming from wind or bad airspeed calibration.
```
```suggestion
 * Excessive airspeed measurements on ground are either caused by wind or bad airspeed calibration.
```

sfuhrer replied:
much better.

RomanBapst commented on src/modules/airspeed_selector/airspeed_selector_params.c:
```diff
@@ -194,3 +194,17 @@ PARAM_DEFINE_INT32(ASPD_FS_T2, 100);
+/**
+ * Maximal airspeed allowed for arming.
+ *
+ * It is not possible to arm the vehicle if the current airspeed estimate (equivalent airspeed) at arming is above this value.
```
```suggestion
 * Deny arming if the current airspeed estimate (equivalent airspeed) is greater than this value.
```

RomanBapst commented on src/modules/airspeed_selector/airspeed_selector_params.c:
```diff
@@ -194,3 +194,17 @@ PARAM_DEFINE_INT32(ASPD_FS_T2, 100);
+ *
+ * @group Airspeed Validator
+ * @unit m/s
+ * @min 0.0
```
@sfuhrer Should we go for something more realistic here, like 0.5m/s? Zero would mean that your sensor needs to return exactly 0 airspeed all the time.

sfuhrer replied:
yes set it to 0.5

sfuhrer commented:
@RomanBapst thanks for the review, I've addressed your comments. 
Do you have an opinion about using the airspeed confidence metric for the preflight checks, resp. of the confidence value of the airspeed overall? @dagar and me discussed it quickly last week, and came to the conclusion that it is probably not worth to keep on dragging that with us, firstly because we don't know of any case where it was detecting a stuck sensor and secondly because it would need a lot of tuning for different sensors (eg for the SDP3X sensors).

Antiheavy commented:
Just posting here to offer my support for this PR.  Looks like a good improvement for those of us who use airspeed.  thanks!

sfuhrer commented:
@Antiheavy do you by chance have any experience with the airspeed confidence indicator? Ever used it/saw it indicating an actual failure?

Antiheavy commented:
> do you by chance have any experience with the airspeed confidence indicator? Ever used it/saw it indicating an actual failure?

Unfortunately I don't think we don't have much experience with "airspeed confidence indicator" yet.  The version of the airspeed failsafe we are using is a pre-v1.9 that we worked with @priseborough  to develop (very similar to what existed in v1.9 with a few minor differences).  I don't think the "confidence indicator" was part of that (was it?).  We have lots of flight data with the airspeed failsafe system using the "load factor ratio" and EKF Innovation test ratio criteria for failure detection.

sfuhrer commented:
> > do you by chance have any experience with the airspeed confidence indicator? Ever used it/saw it indicating an actual failure?
> 
> Unfortunately I don't think we don't have much experience with "airspeed confidence indicator" yet. The version of the airspeed failsafe we are using is a pre-v1.9 that we worked with @priseborough to develop (very similar to what existed in v1.9 with a few minor differences). I don't think the "confidence indicator" was part of that (was it?). We have lots of flight data with the airspeed failsafe system using the "load factor ratio" and EKF Innovation test ratio criteria for failure detection.

@Antiheavy no the airspeed confidence was always (as far as I know) only used to during the preflight checks, not in the in-air airspeed validity check.

RomanBapst commented:
@sfuhrer This is good to go from my side but would be good to get some testing on it.
If other people are going to test it then a test plan might be useful.

LorenzMeier commented:
@sfuhrer Format check failed - do you have the pre-commit hook for format check installed? Run `make check_format` once to get it installed.

sfuhrer commented:
@LorenzMeier yeah I do, but the hook probably doesn't catch formatting issues during a rebase (which is where the format check slipped in I think). I've fixed it now, thanks for pointing it out!

sfuhrer commented:
@RomanBapst I've tested this on my own setups now, both with and without airspeed sensor. No problems seen, so from my point of view we could bring this in.
The only thing I'm still not entirely sure is if we should keep the airspeed confidence metric, as I never yet encountered a situation where it correctly detected a malfunctioning airspeed sensor.

sfuhrer commented:
@dagar as discussed on the devcall this week I removed any dependency of the preflight checks to the airspeed confidence level. As we then don't have any other consumer (beside HIL, see below), I don't see any reason to keep the airspeed confidence calculation and removed that as well from sensors. Do you agree?

sfuhrer commented on src/modules/mavlink/mavlink_high_latency2.cpp:
```diff
@@ -203,10 +203,6 @@ bool MavlinkStreamHighLatency2::write_airspeed(mavlink_high_latency2_t *msg)
-		if (airspeed.confidence < 0.95f) { // the same threshold as for the commander
-			msg->failure_flags |= HL_FAILURE_FLAG_DIFFERENTIAL_PRESSURE;
-		}
-
```
@acfloria as I think you were the one that implemented this: any problem with removing the check for the low confidence here? If yes, then I propose we should subscribe to `airspeed_validated` here, as there you would get notified if the airspeed got invalid (it will set it to NAN). 
Plus, was the airspeed confidence level check working for your setups? Were you able to e.g. detect stuck sensors?

acfloria replied:
If you removed now the confidence and replace it with airspeed_validated it would be nice to also reflect that here.

We never had a faulty airspeed sensor during flights so I cannot report if the confidence was working properly.

sfuhrer replied:
@acfloria like this? https://github.com/PX4/Firmware/pull/14254/commits/324f4fd6150639b6a3c797be2b910a6ec024f89a I'm a HITL novice, so don't know what the goal is that we want to achieve with it exactly. Is the check there just to indicate it if you do not have an airspeed sensor connected to your HITL setup, as part of the preflight checks?

acfloria replied:
Yeah, this seems to be correct. So this message was not used in HIL but when flying with a satellite communication link. Since this message is only sent while flying we did not use it for the preflight checks but as an indication if the airspeed sensor would have been faulty during flight. I think the way you implemented it now it would exactly serve that purpose again.

sfuhrer replied:
Ah of course, now it makes more sense^^ Don't know where I got the idea from that it was HIL stuff.. Okay, thanks for the mini-review!

dagar commented on src/modules/airspeed_selector/airspeed_selector_params.c:
```diff
@@ -194,3 +194,17 @@ PARAM_DEFINE_INT32(ASPD_FS_T2, 100);
+ * @max 30.0
+ * @increment 0.1
+ */
+PARAM_DEFINE_FLOAT(ASPD_MAX_ARM, 4.0f);
```
It's minor, but I think this should be a commander parameter in the current implementation. There are system wide exceptions, but generally parameters shouldn't be used outside of the relevant module and uORB used for anything shared.

sfuhrer replied:
Right, I've changed it, now it is a commander parameter.

sfuhrer commented:
@dagar I've revered the last commit, where I had removed the airspeed confidence completely. Not it's still there, but no longer used inside the preflight checks. Is this now good to go in from your perspective?
